components/brave_rewards/browser/rewards_service_impl.h

```diff
@@ -783,8 +766,12 @@ class RewardsServiceImpl : public RewardsService,
   mojo::AssociatedRemote<bat_ledger::mojom::BatLedger> bat_ledger_;
   mojo::Remote<bat_ledger::mojom::BatLedgerService> bat_ledger_service_;
   const scoped_refptr<base::SequencedTaskRunner> file_task_runner_;
+  base::CancelableTaskTracker cancelable_task_tracker_;
+
+  // Note: Only access |diagnostic_log_| from the file task runner.
+  FileWeakWrapper diagnostic_log_;
```

Comment:
I'm still wondering if this is not thread-safe...

Even though `WeakPtr` can be passed between threads, it's generally not thread-safe to delete an object on a different sequence from which its weak pointers have been dereferenced, because the object destruction could be interleaved with access to the weak pointer on another thread.

I think the object needs to be destroyed over on the file task sequence. If so, then we might need to use an approach like we use for `LedgerDatabase` (i.e. `DeleteSoon(...)`).

Thoughts, @bridiver?

Reply:
@zenparsing correct, this is not safe

Reply:
Would the "right thing" be to use `std::unique_ptr<base::File, base::OnTaskRunnerDeleter>` and pass the raw pointer over to the `...OnTask` functions? ([OnTaskRunnerDeleter](https://source.chromium.org/chromium/chromium/src/+/master:base/sequenced_task_runner.h;l=182;bpv=1;bpt=1?q=ontaskrunnerdeleter&ss=chromium))

Honestly, it seems like the simplest approach would be to just have `std::shared_ptr<base::File>` and pass the `shared_ptr` over to the `...OnTask` functions. The last one to have a shared ptr on any thread (either `RewardsServiceImpl` or one of these `base::OnceCallback` objects) will do the `delete` for us, and IIUC the shared pointer's ref count increment/decrement is thread-safe.

Reply:
@bridiver Recommended checking out `base::FileProxy` which automatically handles the deletion issue (among other functionality). It's very useful but not quite a perfect fit for our use case, so I'm just going to post a deletion task to the file task runner in the Rewards service destructor (which is what `base::FileProxy` is doing under the covers).

Reply:
@zenparsing FWIW we basically should never use `std::shared_ptr`, rather scoped_refptr in rare cases